### PR TITLE
Don't run user function until all UserRRefs in the args are confirmed

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34497 Don't run user function until all UserRRefs in the args are confirmed**
* #34809 Fix Error Hanlding for Local Send
* #34496 Split deserialize from runPythonUdf and remove generatePythonUDFResult

Use a thread_local table to intercept UserRRefs created during user
function args deserialization, and then wait for confirmations of
those UserRRefs before launching the given user function.

Differential Revision: [D20347464](https://our.internmc.facebook.com/intern/diff/D20347464)

### CODE REVIEW

File: torch/csrc/distributed/rpc/rref_impl.h
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/rref_impl.h]: Should we add a comment here that for OwnerRRef it's always confirmed to avoid confusion?

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (rohan-varma) [inline torch/testing/_internal/distributed/rpc/rpc_test.py]: Is it possible to add a test case where a UserRRef is never confirmed and verify the appropriate behavior there? (i.e. stall forever and not run the function, throw an appropriate exception, etc).

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (rohan-varma) [inline torch/csrc/distributed/rpc/rref_context.cpp]: Why do we want to stop recording additional pending users once we've called `waitForThreadLocalPendingUsers`?

**Review by rohan-varma — APPROVED** [review]
Looks good, some minor questions. Although we should also get a review from @pritamdamania87 or @zhaojuanmao.

File: torch/testing/_internal/distributed/rpc/rpc_test.py
Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/rpc_test.py]: I thought about this. Looks like we would need to deplete the threads on the user side or owner side to prevent them either from sending out the RRef request or confirmation. But that threading behavior will change when we introduce async user function. Let me spend some more time to see if it is possible to test it without messing up with threads.

File: torch/csrc/distributed/rpc/rref_context.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/rref_context.cpp]: pending users are added into the context for two reasons:

1. when we call `rpc.remote`, the return value `UserRRef` on the caller will be added to the context.
2. when we forward an RRef to another user, the `UserRRef` on the destination will be added to the context. 

For 1, we don't block as we would like to use that `UserRRef` right away. We only block in 2. So, if we don't close the recording session, there could be more `UserRRef`s added into the `userTable_`. Consider a case where the application runs a user function remotely, and the user function creates a RRef using `rpc.remote`. This RRef should not block the next user function handled by the same RPC thread. 

Thanks for pointing this out, let me add this description to the code.

File: torch/csrc/distributed/rpc/rref_context.h
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/rref_context.h]: nit: PendingRRefs instead of PendingUsers

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Shouldn't this apply to all Messages? If in the future I build a custom message for internal system messages and send an RRef in it, we should still honor the same protocol where we wait for confirmation right?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Should we just do this for all message types and also record for all message types?

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: Can we do this in an async way where we don't block the RPC thread? We can add callbacks on the futures in userTable_ to mark an 'uber' future as complete once done. Then, the 'uber' future can have a callback that actually does all the RPC execution itself. Once the RPC execution is done, we can mark the future that was returned to the RPC layer as complete so that it can respond back to the user.

Given we plan to merge RPC and JIT pools, we can't have blocking calls like these.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Requesting changes since we should avoid blocking the RPC thread waiting for an RRef to be confirmed.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Sure, will make the change.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: You mean this this line of code will appear in every switch-case clause? This cannot directly wrap the `deserializeRequest` function, because Python deserialization occurs within the clause. We probably can add a message layer on top of PythonCall to hold the deserialized py::object. Let me try.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Yes, this should work. let me try.

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (pritamdamania87) [inline torch/csrc/distributed/rpc/request_callback_impl.cpp]: nit: better to do this with RAII.

**Review by pritamdamania87 — APPROVED** [review]

File: torch/csrc/distributed/rpc/request_callback_impl.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/request_callback_impl.cpp]: Will do this in the next PR

**Pull request closed without merging**
